File: Dog.java
/**
 * @author Jonas Andersson Fagerlund joan1043
 */

import java.util.Arrays;

public class Dog{

  private String name;
  private int age;
  private String breed;
  private int weight;
  private double tailLength;
  private Owner owner;

  // konstruktor
  public Dog(String name, String breed, int age, int weight){
    this.name = formatName(name);
    this.breed = formatName(breed);
    this.age = age;
    this.weight = weight;
    setTailLength(age, formatName(breed), weight);
  }

  public void increaseAge(int years) {
    if (years > 0){
      age += years;
      setTailLength(age, breed, weight);
    }
  }

  public String getName() {
    return name;
  }

  public int getAge() {
    return age;
  }

  public String getBreed() {
    return breed;
  }

  public int getWeight() {
    return weight;
  }

  public Owner getOwner(){
      return owner;
  }

  public double getTailLength() {
    return tailLength;
  }

  public boolean haveOwner(){
    return owner != null;
  }

  public void addOwnerToDog(Owner o){
    if (!haveOwner()) {
      owner = o;
      this.owner.addDogToOwner(this);
    }  
  }

   public void removeOwner(){
     if(haveOwner()){
       owner = null;
     }
   }

  public String toString() {
    if(haveOwner()){
      return String.format("<%s %d %s %d tail %s owner=%s>", name, age, breed, weight, tailLength, owner.getName());
    }
    return String.format("<%s %d %s %d tail %s", name, age, breed, weight, tailLength);
  }

  // Metod för att räkna ut tailLength
  private void setTailLength(int age, String breed, double weight) {
    String[] taxBreed = {"Tax", "Dachshund", "Mäyräkoira", "Teckel"};
    final double TAX_TAIL_LENGTH = 3.7;
    if (Arrays.stream(taxBreed).anyMatch(breed::equals)) {
      tailLength = TAX_TAIL_LENGTH;
    } else {
      tailLength = age * weight / 10;
    }
  }

  private String formatName(String str) {
    str = str.trim();
    return str.substring(0, 1).toUpperCase() + str.substring(1).toLowerCase();
  }

}

File: pyDog/dog.py
class Dog:

    def __init__(self, name, breed, age, weight):
        self.__name = name
        self.__breed = breed
        self.__age = age
        self.__weight = weight
        self.__tail_length = self.__calc_tail_length()
        self.__owner = None

    def get_tail_length(self):
        return self.__tail_length

    def get_owner(self):
        return self.__owner

    def get_name(self):
        return self.__name

    def add_owner(self, owner):
        if self.have_owner():
            print('Error: dog already have owner')
        else:
            self.__owner = owner
            owner.add_dog(self)

    def have_owner(self):
        if self.__owner == None:
            return False
        else:
            return True

    def get_age(self):
        return self.__age


    def increase_age(self):
        self.__age += 1

    def __calc_tail_length(self):
        taxes = ['tax', 'dashound']
        if self.__breed.lower() in taxes:
            return 3.7
        else:
            return self.__age * (self.__weight / 10)
        

    def __str__(self):
        if self.have_owner():
            return f'{self.__name}, {self.__tail_length}, {self.__breed}, {self.__age}, {self.__weight}, owner: {self.__owner.get_name()}'
        else:
            return f'{self.__name}, {self.__tail_length}, {self.__breed}, {self.__age}, {self.__weight}'
